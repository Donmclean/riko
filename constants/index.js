export const runCommands = {
    'react-native': ['react-native-ios', 'react-native-launch-android','react-native-android'],
    react: ['react-dev', 'react-prod', 'react-server', 'react-prod-server'],
    electron: ['electron-dev', 'electron-prod', 'electron-server'],
    'node-server': ['node-server-dev', 'node-server-prod']
};