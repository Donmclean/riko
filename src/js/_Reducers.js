import { combineReducers, createStore } from 'redux';

//IMPORT ALL REDUCERS HERE
import { routerReducer } from 'react-router-redux';
import postSelectorReducer from './components/Shared/PostSelector/postSelectorReducer';
import counterReducer from './components/Shared/CounterClicker/counterReducer';

//ADD ALL REDUCERS TO THE 'allReducers' OBJECT HERE
const allReducers = {
    routing: routerReducer,
    postSelectorReducer,
    counterReducer
};

//REDUCERS ARE THEN COMBINED TO BE USED WITH REDUX
const combinedReducers = combineReducers(allReducers);

//THIS OBJECT IS EXPORTED AND CAN/WILL BE USED FOR TESTING PURPOSES
export const mockStoreInitialState = createStore(combinedReducers).getState();

export default combinedReducers;