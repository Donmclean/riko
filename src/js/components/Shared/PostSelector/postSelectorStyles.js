import styled from 'styled-components';

export const PostInput = styled.div``;

export const PostLabel = styled.label`
    display: block;
    color: gray;
    padding-bottom: 5px;
`;

export const PostSelect = styled.select`
    background: gray;
    width: 175px;
    font-size: 20px;
    color: white;
    outline: none;
    border: none;
    padding: 10px;
    text-align-last: center;
    appearance: none;
    cursor: pointer;
`;