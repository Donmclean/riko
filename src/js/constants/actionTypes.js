export const SHOULD_INCREMENT = 'SHOULD_INCREMENT';
export const SHOULD_DECREMENT = 'SHOULD_DECREMENT';
export const INCREMENT = 'INCREMENT';
export const DECREMENT = 'DECREMENT';

export const UPDATE_POSTS = 'UPDATE_POSTS';
export const UPDATE_POST_NUMBER = 'UPDATE_POST_NUMBER';