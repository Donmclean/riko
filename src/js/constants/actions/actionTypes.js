export const INCREMENT = 'INCREMENT';
export const DECREMENT = 'DECREMENT';

export const GET_POSTS = 'GET_POSTS';
export const UPDATE_INPUT_VALUE = 'UPDATE_INPUT_VALUE';