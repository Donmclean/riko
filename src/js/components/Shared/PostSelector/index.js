import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';

//IMPORTANT!: Import actions to dispatch
import getActions from '../../../_ActionCreators';

//Import Styled Components
import { PostSelect, PostLabel, PostInput } from './postSelectorStyles';

class PostSelector extends Component {
    constructor(props) {
        super(props);
        this.actions = getActions(this.props.dispatch);
    }

    handlePostNumberSelect(event) {
        const postNumber = Number(event.target.value.replace(/\D/g, ''));

        this.actions.postSelectorActionCreators.updatePostNumber(postNumber);
        this.actions.postSelectorActionCreators.fetchPosts(postNumber);
    }

    render() {
        //Notice how state is destructured and applied
        const { state } = this.props;
        return (
            <div>
                <h4>Asynchronous Actions</h4>
                <PostInput>
                    <PostLabel>Select Post Number To Fetch</PostLabel>
                    <PostSelect onChange={this.handlePostNumberSelect.bind(this)} value={state.postSelectorReducer.get('postNumber')}>
                        <option disabled="true" defaultValue={0}>{0}</option>
                        {/*GENERATES 100 <option> elements with values of 1 to 100*/}
                        {Array.from(new Array(100).keys()).map((number) => <option key={number + 1} value={number + 1}>{number + 1}</option>)}
                    </PostSelect>
                </PostInput>
            </div>
        );
    }
}

PostSelector.propTypes = {
    state: PropTypes.object.isRequired,
    dispatch: PropTypes.func.isRequired
};

//'state' gets applied to components' props here via 'connect'.
const mapStateToProps = (state) => {
    return {state};
};

export default connect(mapStateToProps)(PostSelector);