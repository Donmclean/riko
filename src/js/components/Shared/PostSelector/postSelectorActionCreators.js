"use strict";
import axios from 'axios';
import * as types from '../../../constants/actionTypes';

export const updatePostNumber = (postNumber) => ({
    type: types.UPDATE_POST_NUMBER,
    data: postNumber
});

export const setPosts = (posts) => ({
    type: types.UPDATE_POSTS,
    data: posts
});

export const fetchPosts = (postId) => {
    return (dispatch) => {
        return axios.get(`https://jsonplaceholder.typicode.com/posts/${postId}`)
            .then((response) => {
                dispatch(setPosts(response.data));
            })
            .catch((error) => {
                return new Error(error);
            });
    };
};