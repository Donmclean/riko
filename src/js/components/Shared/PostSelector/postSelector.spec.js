import React from 'react';
import _ from 'lodash';
import * as postSelectorActions from './postSelectorActionCreators';
import * as types from '../../../constants/actionTypes';

//For testing components
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
const middlewares = [ thunk ];
const mockStore = configureStore(middlewares);

// Useful for not making direct api requests when testing
// import nock from 'nock';

//This contains all reducers and their 'initialState' ready to be placed in 'mockStore' for testing
//This should be ideal for testing dummy components
import { mockStoreInitialState } from '../../../_Reducers';

//Component to be Tested
import PostSelector from './index';

describe('postSelector Test Suite', () => {
    let store;

    beforeEach(() => {
        //Initialize mockStore with initial state of the reducers
        store = mockStore(mockStoreInitialState);
    });

    //Action Creator Specific Tests would go here
    describe('postSelector Actions', () => {
        it(`setPosts should create an ${types.UPDATE_POSTS} action`, () => {
            //OLD WAY
            // const expectedAction = { type: types.UPDATE_POSTS };
            // expect(postSelectorActions.setPosts()).toEqual(expectedAction);

            //NEW WAY!!!
            expect(postSelectorActions.setPosts()).toMatchSnapshot();
        });

        it(`updatePostNumber should create an ${types.UPDATE_POST_NUMBER} action`, () => {
            const postNumber = 10;

            const dispatch = postSelectorActions.updatePostNumber(postNumber);

            //OLD WAY
            // const expectedDispatch = { type: types.UPDATE_POST_NUMBER, data: 10 };
            // expect(dispatch).toEqual(expectedDispatch);

            //NEW WAY!!!
            expect(dispatch).toMatchSnapshot();
        });

        it('fetchPosts should return a dispatch function', () => {
            const postId = 1;
            const dispatch = postSelectorActions.fetchPosts(postId);
            expect(_.isFunction(dispatch)).toEqual(true);
        });
    });

    describe('postSelector Dispatch Actions', () => {
        it(`fetchPosts creates ${types.UPDATE_POSTS} when fetching posts has been done`, () => {
            const postId = 1;

            const store = mockStore();

            return store.dispatch(postSelectorActions.fetchPosts(postId))
                .then(() => { // return of async actions
                    expect(store.getActions()).toMatchSnapshot();
                });
        });

        it('fetchPosts should return empty array upon error occurring when fetching invalid post number or (404)', () => {
            const postId = 0;
            const store = mockStore();

            return store.dispatch(postSelectorActions.fetchPosts(postId))
                .then(() => { // return of async actions
                    expect(store.getActions()).toEqual([]);
                    expect(_.isEmpty(store.getActions())).toEqual(true);
                });
        });
    });

    describe('PostSelector Component', () => {
        it('renders <PostSelector /> component successfully in to the DOM with required props & state', () => {

            //Verify props
            const wrapper = shallow(<PostSelector store={store}/>);
            expect(wrapper.prop('store')).toEqual(store);

            //Verify that initial State of the store is rendered correctly
            expect(wrapper.state().storeState).toEqual(store.getState());

            //Render Component Tree
            const component = renderer.create(
                <PostSelector store={store}/>
            );

            //Verify Component Snapshot
            //call component to JSON before matching snapshot
            expect(component.toJSON()).toMatchSnapshot();
        });
    });
});