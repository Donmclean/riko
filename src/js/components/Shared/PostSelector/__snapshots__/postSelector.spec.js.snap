exports[`postSelector Test Suite PostSelector Component renders <PostSelector /> component successfully in to the DOM with required props & state 1`] = `
<div>
  <h4>
    Asynchronous Actions
  </h4>
  <div
    className="async-input">
    <label>
      Select Post Number To Fetch
    </label>
    <select
      onChange={[Function]}
      value={0}>
      <option
        defaultValue={0}
        disabled="true">
        0
      </option>
      <option
        value={1}>
        1
      </option>
      <option
        value={2}>
        2
      </option>
      <option
        value={3}>
        3
      </option>
      <option
        value={4}>
        4
      </option>
      <option
        value={5}>
        5
      </option>
      <option
        value={6}>
        6
      </option>
      <option
        value={7}>
        7
      </option>
      <option
        value={8}>
        8
      </option>
      <option
        value={9}>
        9
      </option>
      <option
        value={10}>
        10
      </option>
      <option
        value={11}>
        11
      </option>
      <option
        value={12}>
        12
      </option>
      <option
        value={13}>
        13
      </option>
      <option
        value={14}>
        14
      </option>
      <option
        value={15}>
        15
      </option>
      <option
        value={16}>
        16
      </option>
      <option
        value={17}>
        17
      </option>
      <option
        value={18}>
        18
      </option>
      <option
        value={19}>
        19
      </option>
      <option
        value={20}>
        20
      </option>
      <option
        value={21}>
        21
      </option>
      <option
        value={22}>
        22
      </option>
      <option
        value={23}>
        23
      </option>
      <option
        value={24}>
        24
      </option>
      <option
        value={25}>
        25
      </option>
      <option
        value={26}>
        26
      </option>
      <option
        value={27}>
        27
      </option>
      <option
        value={28}>
        28
      </option>
      <option
        value={29}>
        29
      </option>
      <option
        value={30}>
        30
      </option>
      <option
        value={31}>
        31
      </option>
      <option
        value={32}>
        32
      </option>
      <option
        value={33}>
        33
      </option>
      <option
        value={34}>
        34
      </option>
      <option
        value={35}>
        35
      </option>
      <option
        value={36}>
        36
      </option>
      <option
        value={37}>
        37
      </option>
      <option
        value={38}>
        38
      </option>
      <option
        value={39}>
        39
      </option>
      <option
        value={40}>
        40
      </option>
      <option
        value={41}>
        41
      </option>
      <option
        value={42}>
        42
      </option>
      <option
        value={43}>
        43
      </option>
      <option
        value={44}>
        44
      </option>
      <option
        value={45}>
        45
      </option>
      <option
        value={46}>
        46
      </option>
      <option
        value={47}>
        47
      </option>
      <option
        value={48}>
        48
      </option>
      <option
        value={49}>
        49
      </option>
      <option
        value={50}>
        50
      </option>
      <option
        value={51}>
        51
      </option>
      <option
        value={52}>
        52
      </option>
      <option
        value={53}>
        53
      </option>
      <option
        value={54}>
        54
      </option>
      <option
        value={55}>
        55
      </option>
      <option
        value={56}>
        56
      </option>
      <option
        value={57}>
        57
      </option>
      <option
        value={58}>
        58
      </option>
      <option
        value={59}>
        59
      </option>
      <option
        value={60}>
        60
      </option>
      <option
        value={61}>
        61
      </option>
      <option
        value={62}>
        62
      </option>
      <option
        value={63}>
        63
      </option>
      <option
        value={64}>
        64
      </option>
      <option
        value={65}>
        65
      </option>
      <option
        value={66}>
        66
      </option>
      <option
        value={67}>
        67
      </option>
      <option
        value={68}>
        68
      </option>
      <option
        value={69}>
        69
      </option>
      <option
        value={70}>
        70
      </option>
      <option
        value={71}>
        71
      </option>
      <option
        value={72}>
        72
      </option>
      <option
        value={73}>
        73
      </option>
      <option
        value={74}>
        74
      </option>
      <option
        value={75}>
        75
      </option>
      <option
        value={76}>
        76
      </option>
      <option
        value={77}>
        77
      </option>
      <option
        value={78}>
        78
      </option>
      <option
        value={79}>
        79
      </option>
      <option
        value={80}>
        80
      </option>
      <option
        value={81}>
        81
      </option>
      <option
        value={82}>
        82
      </option>
      <option
        value={83}>
        83
      </option>
      <option
        value={84}>
        84
      </option>
      <option
        value={85}>
        85
      </option>
      <option
        value={86}>
        86
      </option>
      <option
        value={87}>
        87
      </option>
      <option
        value={88}>
        88
      </option>
      <option
        value={89}>
        89
      </option>
      <option
        value={90}>
        90
      </option>
      <option
        value={91}>
        91
      </option>
      <option
        value={92}>
        92
      </option>
      <option
        value={93}>
        93
      </option>
      <option
        value={94}>
        94
      </option>
      <option
        value={95}>
        95
      </option>
      <option
        value={96}>
        96
      </option>
      <option
        value={97}>
        97
      </option>
      <option
        value={98}>
        98
      </option>
      <option
        value={99}>
        99
      </option>
      <option
        value={100}>
        100
      </option>
    </select>
  </div>
</div>
`;

exports[`postSelector Test Suite postSelector Actions setPosts should create an UPDATE_POSTS action 1`] = `
Object {
  "data": undefined,
  "type": "UPDATE_POSTS",
}
`;

exports[`postSelector Test Suite postSelector Actions updatePostNumber should create an UPDATE_POST_NUMBER action 1`] = `
Object {
  "data": 10,
  "type": "UPDATE_POST_NUMBER",
}
`;

exports[`postSelector Test Suite postSelector Dispatch Actions fetchPosts creates UPDATE_POSTS when fetching posts has been done 1`] = `
Array [
  Object {
    "data": Object {
      "body": "quia et suscipit
suscipit recusandae consequuntur expedita et cum
reprehenderit molestiae ut ut quas totam
nostrum rerum est autem sunt rem eveniet architecto",
      "id": 1,
      "title": "sunt aut facere repellat provident occaecati excepturi optio reprehenderit",
      "userId": 1,
    },
    "type": "UPDATE_POSTS",
  },
]
`;
